KEP-5517: Introduce DRA for Node Allocatable Resources KEP

pravk03 opened the pull request:
<!-- 
	Please use the following format when naming your PR
	< Issue Number >:< Issue Description >
	e.g. KEP-000: adding beta graduation criteria
	
	Avoid using phrases like `fixes #NNNN` in the description
	unless the pull request is to change the KEP status to 
	implemented or KEP has been deprecated.
-->

<!-- short description of work done in PR e.g. updating milestone, adding new KEP, adding test requirements… -->  
- One-line PR description:  Support native resource management through DRA.

<!-- link to the k/enhancements issue -->
- Issue link: https://github.com/kubernetes/enhancements/issues/5517

<!-- other comments or additional information -->
- Other comments:

pravk03 committed:
KEP-5517: Introduce DRA for Native Resources KEP

ffromani commented:
/cc

lmktfy commented on keps/sig-node/5517-dra-native-resources/kep.yaml:
```diff
@@ -0,0 +1,37 @@
+title: DRA for Native Resources
```
The issue title is: "DRA: Native Resource Requests"

Let's match that, or change the issue title to match this.

ffromani commented on keps/sig-scheduling/5517-dra-native-resources/README.md:
```diff
@@ -0,0 +1,1717 @@
+
+Dynamic Resource Allocation (DRA) provides a powerful framework for managing specialized hardware resources such as GPUs, FPGAs, and high-performance network interfaces. It also enables fine-grained management of native resources like CPU and Memory, for example, through the [dra-driver-cpu](https://github.com/kubernetes-sigs/dra-driver-cpu). However, when a native resource is managed via DRA, a  ...
+
+Additionally, specialized resources like accelerators have implicit dependency on native resources like CPU or Hugepages for the application to interact with it. Currently, users must manually research and declare these auxiliary native resource requirements, typically as additional requests in the PodSpec. This process is error-prone and adds complexity to workload configuration.
```
it may be worth to spell out explicitly that, even if users endure the error-prone mechanism, the key point above about the mismatch between core resources and DRA prevents to incorporate these requirements in a cohesive way.
If my accelerator requires some hugepages NUMA-aligned to the PCIe bus my accelerator resides on, there's no way to express this constraint with the DRA APIs - and possibly in general due the double accounting.

ffromani commented on keps/sig-scheduling/5517-dra-native-resources/README.md:
```diff
@@ -0,0 +1,1717 @@
+
+### Background
+
+To understand the proposed solution, it is essential to first understand how kube-scheduler currently manage standard resource requests and DRA ResourceClaims.
```
*perhaps* - just a wild thought - the solution can "just" be to introduce helpers to implement the joint accounting, and consume them first in the scheduler and then anywhere else needed?

pravk03 replied:
Covered this in the "Resource Calculation" section.

ffromani commented on keps/sig-scheduling/5517-dra-native-resources/README.md:
```diff
@@ -0,0 +1,1717 @@
+  ClaimInfo ObjectReference
+  // Containers lists the names of all containers in this pod that are
+  // sharing the allocation from this claim.
+  Containers []string
```
the fact we need this field makes me wonder if the concept is better expressed per-container, kinda like `Pod.Status.ContainerStatuses`

ffromani commented on keps/sig-scheduling/5517-dra-native-resources/README.md:
```diff
@@ -0,0 +1,1717 @@
+    *   **DynamicResources Plugin**:Executes its standard PreFilter checks on all ResourceClaim objects, including validation of claim and class existence. For claims involving native resources, it additionally validates by checking if aggregation policies of different claims are compatible. Any failure here makes the pod unschedulable on any node.
+
+2.  **Filter Stage:** This stage performs the node-level checks to determine if a pod fits on a specific node.
+    *   **NodeResourcesFit Plugin:** This plugin runs first. If it determined in PreFilter that DRA manages native resources for this pod, it *skips* all standard native resource capacity checks (CPU, Memory, HugePages, etc.), effectively deferring the responsibility to the `DynamicResources` plugin.
```
the assumption `NodeResourcesFit` runs first needs either a more formal contract (e.g. API guarantee) or should be removed IMO (again pending sig-scheduling review). OTOH it seems to me that in both `Filter` and `PreFilter` stage we don't really have any ordering constraints. If the `NodeResourcesFit` just shortcircutis to NOP, nothing changes in the proposed flow.

ffromani commented on keps/sig-scheduling/5517-dra-native-resources/README.md:
```diff
@@ -0,0 +1,1717 @@
+As mentioned in the Non-Goals, full Kubelet awareness of DRA native resources for QoS classification and cgroup management is not in scope for Alpha. Future work will involve:
+
+*   Updating Kubelet's QoS class calculation to include native resources from `pod.Status.NativeResourceClaimStatus`.
+*   Ensuring Kubelet's cgroup manager correctly configures CPU and Memory limits/shares based on the sum of PodSpec requests and DRA-provided native resources.
```
where's the right place to discuss _potential future_ responsability transfer from the kubelet to a driver to manage thw cgroup layout? this KEP in a future cycle or another place (which?)

ffromani reviewed (commented):
Thanks for this work! I claim familiarity with this proposal, having reviewed earlier design doc drafts and partecipated in many long conversations. Yet, it's a very complex and hard topic, so this is my initial partial review

pravk03 commented on keps/sig-scheduling/5517-dra-native-resources/README.md:
```diff
@@ -0,0 +1,1717 @@
+
+## Design Details
+
+The proposal here is to enhance the kube-scheduler to implement a **"Unified Accounting"** model for native resources requested through the standard pod Spec or through Dynamic Resource Allocation (DRA) claims. This involves modifications in `NodeResourcesFit` and `DynamicResources` plugins in how they track resource usage on the node. This also includes updates to the DRA API for drivers to decl ...
```
After exploring a bit more, the plugin order is not a strict requirement. The delegation mechanism should work irrespective of the order. Updated this section.

pravk03 commented:
/cc @dchen1107

liggitt commented on keps/sig-scheduling/5517-dra-native-resources/README.md:
```diff
@@ -0,0 +1,1717 @@
+  * Via the DRA driver in `ResourceSlice` objects.
+
+* **Dual Consumption:** Pods can consume this CPU capacity in two different ways:  
+  * Via `pod.spec.resources`, which is accounted for by the `NodeResourcesFit` scheduler plugin.  
```
(and in `pod.spec.{containers,initContainers}[*].resources`, `pod.spec.overhead`, I think)

pravk03 commented on keps/sig-scheduling/5517-dra-native-resources/README.md:
```diff
@@ -0,0 +1,1717 @@
+  // PolicyAddPerReference means the DRA claim is added for each container referencing the claim.
+  PolicyAddPerReference NativeResourceAccountingPolicy = "AddPerReference"
+  // PolicyConsumeFrom means the container's standard requests are drawn from the pool defined by the DRA claim.
+  PolicyConsumeFrom      NativeResourceAccountingPolicy = "ConsumeFrom"
```
Yes, `ConsumeFrom` policy is different as it changes the source of resource consumption. Instead of the standard request claims consuming from the node allocatable, it would be consuming from a resource claim. The claim request is treated as a reservation pool and is allocated when the first pod references it and subsequent pod references draw from this pool.

swatisehgal commented on keps/sig-scheduling/5517-dra-native-resources/README.md:
```diff
@@ -0,0 +1,1759 @@
+* To move all resource management logic into the DRA driver. The Kubelet will remain the primary agent for cgroup management and QoS enforcement, ensuring that the benefits of its existing stability and lifecycle management features are preserved.  
+* To replace the standard `pod.spec.containers.resources` API for requesting native resources. This KEP aims to enhances the system by adding a clear path for native resource requests via DRA while ensuring it works coherently with the
+  existing PodSpec-based requests.
+* Changes to the Kubelet for QoS classification, cgroup management, and eviction logic based on DRA native resource allocations are not in scope for the Alpha release of this KEP.
```
If kubelet QoS integration is deferred to future work, what is the expected behavior for QoS classification and eviction during the Alpha stage? Specifically, how will we ensure that Pods running infrastructure DRA drivers are not evicted by other resource-hungry workloads on the same node?

swatisehgal reviewed (commented):
Partial Review: From what I have read so far, my broader concern is that QoS integration is being deferred to future work. This leaves a significant gap in how Pods consuming DRA-managed native resources are handled today. Without clear QoS accounting, it’s unclear how the scheduler and kubelet will classify these Pods, and whether critical infrastructure Pods using DRA claims could be inadvertently treated as lower QoS and evicted under resource pressure. 

This uncertainty affects both the safety of deployments and the realistic usability of the Alpha feature. It would be valuable to explicitly document interim behavior, potential risks, and mitigation strategies for QoS and eviction during Alpha.

pohly commented on keps/sig-scheduling/5517-dra-native-resources/README.md:
```diff
@@ -0,0 +1,1579 @@
+type NativeResourceQuantity struct {
+    // PerInstanceQuantity: Each allocated device instance contributes this Quantity to the native resource.
+    // Used when devices in the ResourceSlice represent discrete units of the native resource.
+    // +optional
```
```suggestion
    // +optional
    // +oneOf=Quantity
```

pravk03 commented on keps/sig-scheduling/5517-dra-native-resources/README.md:
```diff
@@ -0,0 +1,1579 @@
+
+## Summary
+
+This KEP proposes a solution for managing native resources like CPU, Memory and Hugepages with Dynamic Resource Allocation (DRA). Currently, when native resources are managed via DRA, there is a fundamental disconnect across the control plane and the Node. In the scheduler, having two independent accounting systems (one for standard resources, one for DRA) which are managing the same underlying r ...
```
Ack. Sorry for missing this. 

Will do this for newer sections and in sections where I would not lose any comments.

pravk03 commented on keps/sig-scheduling/5517-dra-native-resources/README.md:
```diff
@@ -0,0 +1,1579 @@
+*   `DeviceClassSpec.NativeResourcePolicies`: This field allows administrators to specify default accounting policies for native resources for devices matching this class. This can be left empty if the `AccountingPolicy` is provided in the `Device` objects. The key of this map is the native resource name (e.g., `cpu`, `memory`, `hugepages-1Gi`)
+    *   **AccountingPolicy:** Defines how the native resource quantity from the DRA claim is accounted for along with any requests for the same resource made in the container's `resources.requests` in the Pod Spec.
+        1.   **AddPerClaim:** The quantity from the DRA claim is added to the standard request for the resource in the pod spec. If the claim is shared by multiple containers in the same pod, the request in the claim is added once to the Pod's total requests for this resource.
+        2.   **AddPerReference:** The quantity from the DRA claim is added to the standard request for the resource in the pod spec. If the claim is shared by multiple containers in the same pod, the request in the claim is added once every time the claim is referenced.
```
Added the description along with the API definition and removed the text outside.

pravk03 commented on keps/sig-scheduling/5517-dra-native-resources/README.md:
```diff
@@ -0,0 +1,1579 @@
+    *   **DynamicResources Plugin:** The plugin updates the `ResourceClaim.Status` to reflect the allocated devices. It also patches the `Pod.Status` to add the `NativeResourceClaimStatus` field. This new field contains information about the native resources being provided via DRA, including the quantities, the `AccountingPolicy` applied, and which containers reference the claim. This makes the D ...
+
+4.  **Bind Stage (Framework Cache Update):** This is the final step where the scheduler records the pod's resource consumption on the node in the `NodeInfo` cache. The core logic in the scheduler framework is enhanced to use `pod.Status.NativeResourceClaimStatus` and new fields in `NodeInfo` to track DRA claim states.
+    *  The scheduler framework's core logic for updating the `NodeInfo.Requested` is enhanced. When a pod is bound ([`NodeInfo.update`](https://github.com/kubernetes/kubernetes/blob/4925c6bea44efd05082cbe03d02409e0e7201252/pkg/scheduler/framework/types.go#L425) method), the framework reads `pod.Status.NativeResourceClaimStatus` and combines that with the standard request to determine the effectiv ...
```
Thanks for raising this. This is a valid concern and I am looking into it. 

IIUC, [AssumePod](https://github.com/kubernetes/kubernetes/blob/4b5e40aacd021aa1485d29c4965893b3de017697/pkg/scheduler/backend/cache/cache.go#L360) is when the current allocation decision gets committed to node info and this gets picked up during subsequent pod allocations. We would likely need to persist the native DRA claim allocation information in cycle state and pass it to this function.

pravk03 commented on keps/sig-scheduling/5517-dra-native-resources/README.md:
```diff
@@ -0,0 +1,1579 @@
+        *   The results of the native resource calculations (quantities and policies per claim) are stored in the internal cycle state for use in the PreBind stage to update the PodStatus.
+
+3.  **PreBind Stage:** This stage performs actions right before the pod is immutably bound to the node.
+    *   **DynamicResources Plugin:** The plugin updates the `ResourceClaim.Status` to reflect the allocated devices. It also patches the `Pod.Status` to add the `NativeResourceClaimStatus` field. This new field contains information about the native resources being provided via DRA, including the quantities, the `AccountingPolicy` applied, and which containers reference the claim. This makes the D ...
```
ACK. I will update this section along with the changes needed to resolve your other comment (Reserve/Bind race).

pravk03 commented on keps/sig-scheduling/5517-dra-native-resources/README.md:
```diff
@@ -0,0 +1,1579 @@
+// NativeResourceQuantity defines the method to identify how we obtain native resource quantity from the Claim.
+// Only one of PerInstanceQuantity or Capacity must be specified.
+type NativeResourceQuantity struct {
+    // PerInstanceQuantity: Each allocated device instance contributes this Quantity to the native resource.
```
Updated the description to add more clarity.

`PerInstanceQuantity` specifies a fixed quantity for the allocated device. 
For eg: A GPU device that is allocated to the claim also request additional 2 GB  (fixed)of memory.

`Capacity` specifies the key that ties the capacity in the slice to what is consumed in the claim.
 For example: if dra.example.com/memory is expressed as a capacity in the resource slice, we add `nativeResourceMappings["memory"].quantityFrom.capacity: "dra.example.com/memory"`. The actual quantity that the claim adds to the pod's native resource footprint is then determined by the scheduler based on the quantity requested in the claim.

pravk03 commented on keps/sig-scheduling/5517-dra-native-resources/README.md:
```diff
@@ -0,0 +1,1579 @@
+  PolicyAddPerClaim      NativeResourceAccountingPolicy = "AddPerClaim"
+  // PolicyAddPerReference means the DRA claim is added to the standard request.  If multiple containers reference the claim, the claim request is added once per each reference.
+  PolicyAddPerReference NativeResourceAccountingPolicy = "AddPerReference"
+)
```
In terms of the kube-scheduler accounting for these two policies, the logic remains the same whether the claim is shared between containers of the same pod or across different pods.

Updated the description to include this.

pravk03 commented on keps/sig-scheduling/5517-dra-native-resources/README.md:
```diff
@@ -0,0 +1,1593 @@
+
+The core function `[resource.PodRequests](https://github.com/kubernetes/kubernetes/blob/4925c6bea44efd05082cbe03d02409e0e7201252/staging/src/k8s.io/component-helpers/resource/helpers.go#L149)` (from `k8s.io/component-helpers/resource`) is the standard API used in several components, including the Kubelet's admission control and the scheduler's `NodeResourcesFit` plugin to determine the aggregate  ...
+
+To support DRA native resources, the `resource.PodRequests` function would be enhanced. When the `DRANativeResources` feature is enabled, this function now inspects the `pod.Status.NativeResourceClaimStatus`. This field is populated by the `DynamicResources` scheduler plugin and contains the quantities of native resources allocated to the pod via its DRA claims. This enhancement ensures that both ...
```
No, the Kubelet does not update `node.Status.Allocatable` based on these calculations as it represents static node capacity (after deducting kube-reserved and system-reserved). The scheduler tracks the available capacity (which is based on all the pods running on this node) in its `NodeInfo` cache. 

> Would pod requests and limits be ignored if pod requests native resources through DRA?

For alpha, the supported policies (`AddPerClaim`, `AddPerReference`) are additive to what is requested in the pod spec.

pravk03 commented on keps/sig-scheduling/5517-dra-native-resources/README.md:
```diff
@@ -0,0 +1,1717 @@
+5.  A container can use a claim with `PolicyOverride` for a native resource (e.g., from a CPU DRA driver) to set its base request, while simultaneously using other claims for the same native resource with `PolicyAddPerClaim` or `PolicyAddPerReference` (e.g., from a GPU driver for auxiliary CPU). The scheduler will sum the overridden value with rest of the additive policies while accounting for no ...
+6.  A container can use a claim with `PolicyConsumeFrom` for a native resource to set its base request, while  using other claims for the same native resource with `PolicyAddPerClaim` or `PolicyAddPerReference` (e.g., from a GPU driver for auxiliary CPU). The container's `resources.requests` are still drawn from the `ConsumeFrom` pool and the `PolicyAddPerClaim`/`PolicyAddPerReference` are accoun ...
+
+This validation occurs within the `Filter` stage of the `DynamicResources` plugin for each node being considered. If any of these rules are violated, the pod is deemed to have an invalid DRA resource configuration, and the `Filter` stage will fail for that node. This will result in the pod being marked as **unschedulable** in the current scheduling cycle.
```
With the updated scope to only include additive policies, we do not need this.

bart0sh commented on keps/sig-scheduling/5517-dra-native-resources/README.md:
```diff
@@ -0,0 +1,1877 @@
+*   Ensuring Kubelet's cgroup manager correctly configures CPU and Memory limits/shares based on the sum of PodSpec 
+    requests and DRA-provided native resources.
+*   Aligning eviction thresholds with the true resource footprint, including DRA.
+
```
Scoring and Quota support should be also mentioned somewhere. I guess they're not in the Alpha scope, but should probably be done for Beta.

pravk03 commented:
/assign @johnbelamaric , @dom4ha , @kad, @ffromani

macsko commented on keps/sig-scheduling/5517-dra-native-resources/README.md:
```diff
@@ -0,0 +1,1877 @@
+-    The Kubelet's admission handler is updated to consider native resource claims in `Pod.Status`.
+-   All unit and integration tests outlined in the Test Plan are implemented and verified.
+
+#### Beta
```
I think we would need a proper scoring in kube-scheduler for beta. The user should expect similar scheduling outcome, regardless of kubelet/DRA-based node resources representation

macsko commented on keps/sig-scheduling/5517-dra-native-resources/README.md:
```diff
@@ -0,0 +1,1877 @@
+the resources added to the NodeInfo cache from `Pod.Status.NativeResourceClaimStatus` would not be
+subtracted when the pod is removed. This would result in the scheduler cache overestimating resource usage
+on the node, potentially preventing new pods from scheduling. This inconsistent state would persist until
+all the pods with native resource claims are restarted or the kube-scheduler is restarted. 
```
Disabling the feature gate in kube-scheduler requires restarting it. This means that the node requests will be recalculated on startup, without taking DRA native resources into consideration. Pod removal won't cause any inconsistency.

pravk03 commented on keps/sig-scheduling/5517-dra-native-resources/README.md:
```diff
@@ -0,0 +1,1877 @@
+    API server) to succeed. Without an "assume" step, the scheduler might try to place other pods on the same node using
+    stale resource information, potentially leading to oversubscription. The Assume phase "reserves" the resources in the
+    scheduler's in-memory cache immediately.
+    *   The scheduler framework reads `PodNativeResourceClaimStatus` from `CycleState`.
```
Updated this section. 

I have started exploring the solution in https://github.com/kubernetes/kubernetes/pull/136725. This is an early draft.

dom4ha commented on keps/sig-scheduling/5517-dra-native-resources/README.md:
```diff
@@ -0,0 +1,1889 @@
+#### Unreferenced Claims
+
+If a `ResourceClaim` is listed in `pod.Spec.ResourceClaims` but not referenced by any container in `pod.Spec.Containers[*].Resources.Claims`.
+The resources associated with this claim ARE still accounted for against the node's capacity. The DRA allocator reserves the devices for the pod,
```
What about in case of "AddPerReference" accounting policy?

pravk03 replied:
> What about in case of "AddPerReference" accounting policy? 

Updated this section. In the case of unreferenced claims, we should account for it once when we update `NodeInfo`  irrespective of the accounting policy.

dom4ha reviewed (commented):
Have you considered defining some sort of framework level consumable counter that different plugins could consume capacity from? DRA devices could use this counter using Consumable Capacity feature and NodeResources plugin would be extended somehow to use it as well.

It would be a bit cleaner conceptually, but we'd introduce quite significant framework change, which may be harder to get through the process.

pravk03 commented:
Summary of modifications made KEP based on discussions. This should help reduce the scope and implementation complexity on the scheduler side.

1. Removed `AccountingPolicy` from `Device` and `DeviceClass` APIs. For the Alpha scope, we will support a single accounting policy where claim requests is allocated once and is additive to the pod's standard resource requests. The `AccountingPolicy` field along with supported policies has been moved to the "Future Enhancements" section for reconsideration in Alpha 2/Beta. 
2.   Sharing native resource claims between pods is excluded from alpha. The `DynamicResources` plugin will be updated to reject any pod that references a claim already in use by an existing pod. This reduces implementation complexity and allows for more discussion on node-level enforcement and cgroup management. 
3.  Pod Level Resources, if specified will take precedence and determine the overall resource footprint. Added a validation step in the `DynamicResources` plugin to ensure that container requests (including native resource claims) do not exceed pod-level requests. 

cc  @johnbelamaric @dchen1107 @SergeyKanzhelev @dom4ha

johnbelamaric reviewed (commented):
PRR is fine for alpha

SergeyKanzhelev commented:
The new scope seems more reasonable. I am still very un-easy on a few topics:

1. Can we GA the Native Resources DRA plugin so people can start adopting it (before we fully GA-ed this KEP)? If so - we may want to define a migration plan or just say that it will be a different DRA plugin once this KEP is GA.
2. Are there real world frameworks or users who lined up to try the fungibility between GPU and CPU? If we have nobody confirming that this is "likely" the way to go - it feels like we are implementing an expensive solution to the problem nobody is solving right now. Will we have enough feedback after alpha to decide on the path forward?

pravk03 commented on keps/sig-scheduling/5517-dra-native-resources/README.md:
```diff
@@ -0,0 +1,1978 @@
+-   The Kubelet's admission handler is updated to consider native resource claims in `Pod.Status`.
+-   All unit and integration tests outlined in the Test Plan are implemented and verified.
+
+#### Alpha2 / Beta
```
Do you mean populating the claim information in pod status rather than ResourceClaim status ?. Or the general idea of having kubelet having to look outside of pod.spec ?

ffromani commented:
> > 1. Can we GA the Native Resources DRA plugin so people can start adopting it (before we fully GA-ed this KEP)? If so - we may want to define a migration plan or just say that it will be a different DRA plugin once this KEP is GA.
> 
> I think the answer depends heavily on the specific plugin implementation. Specifically thinking of [dra-driver-cpu](https://github.com/kubernetes-sigs/dra-driver-cpu), we can support adoption now using existing workarounds, such as mirroring claim requests in the Pod Spec. I view having Pod Level Resources define the overall shape (including claims) as one clean path for migrating to this KEP, but this definitely needs more thought. I am not sure if the same approach works for Memory or other drivers. @ffromani, since you are working on both CPU and Memory drivers, you may have any thoughts here?

Like you, I tend to believe that combining with Pod Level Resources should work, but we need more data here.

I see the point about experimenting with drivers before adding important changes into core kubernetes. I'm not against it, even if I see clearly the value in this KEP. The issues with this approach (try external drivers first() boils down to chicken/egg problems:
- the amount of workarounds needed is likely to limit significantly the UX, then the adoption, which provides us too little data
- what drivers can actually do is limited, which in turn constraints negatively the adoption, which circles back to too little data.

What I'm trying to say is that I see the value in experimenting out of tree and I personally would encourage that but without _some_ core kubernetes changes, be them in the scheduler, in the kubelet or likely both, what drivers for core resources can do is very limited, and that hampers early adoption.
